x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/RBACEngine.java

```diff
@@ -132,6 +135,50 @@ private void getRoles(User user, Authentication authentication, ActionListener<R
         rolesStore.getRoles(user, authentication, listener);
     }
 
+    @Override
+    public boolean isChildActionAuthorizedByParent(RequestInfo requestInfo, String parentAction, IndicesAccessControl parentAccessControl) {
+        final String childAction = requestInfo.getAction();
+        if (IndexPrivilege.CREATE_INDEX_MATCHER.test(childAction) || childAction.equals(TransportShardBulkAction.ACTION_NAME)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (isScrollRelatedAction(childAction) || isAsyncRelatedAction(childAction)) {
+            // These need special handling, so don't short circuit
+            return false;
+        }
+        if (childAction.startsWith(parentAction) == false) {
```

Comment:
Should we be more conservative and have `childAction.startsWith(parentAction + "[")`. Otherwise it is possible for an action to be child of itself, which I don't think makes sense in the known context.

Also, it is not recommended for end users to grant privileges using raw action names. Now it absolutely does not make sense if child action names are used. Should we add some warning or throwing error for such usages when we can detect them? It can be a separate PR.

Reply:
I don't really follow.
`Privilege.actionToPattern` always adds a `*` on the end of any action name that is granted, so anything with that prefix is considered a child for the purposes of privilege checking.